Add a consistent rendering protocol.

pelme opened the pull request:
This change provides a consistent API to render a htpy object as HTML or
iterate over it.

This commit introduces the iter_chunks() method which is identical with
`__iter__()` but with a better name.

With the introduction of Fragment, this commit makes render_node and
iter_node redundant.

This commit deprecates render_node, iter_node and direct iteration of elements.

More info: https://github.com/pelme/htpy/pull/86#discussion_r1986098069

davepeck commented:
I've been playing with these changes (well, after fixing that one lint issue) and: so far, so good. All my "template" functions return `h.Renderable` now.

It also caused me to slightly update my `@with_children` decorator, which I still use to make regular functions support `__getitem__` and look a bit more like the rest of the flow:

```python
@dataclass(frozen=True)
class with_children[C, R: h.Renderable, **P]:
    """Wrap a function to make it look more like an htpy.Element."""

    _f: t.Callable[t.Concatenate[C, P], R]
    _args: tuple[t.Any, ...] = field(default_factory=tuple)
    _kwargs: t.Mapping[str, t.Any] = field(default_factory=dict)

    def __getitem__(self, children: C) -> R:
        """Render the component with the given children."""
        return self._f(children, *self._args, **self._kwargs)

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> t.Self:
        """Return a new instance of the class with the given arguments."""
        return replace(self, _args=args, _kwargs=kwargs)

    def __str__(self) -> str:
        """Return the name of the function being wrapped."""
        # CONSIDER: alternatively, invoke the function here? If the function
        # provides a default value for its arguments, it'll work; otherwise,
        # it will blow up... which might be a good thing?
        return f"with_children[{self._f.__name__}]"
```

pelme commented:
@davepeck awesome, thanks for the update! pushed an updated version that fixes the lint problem.

I renamed `stream_chunks()` to `iter_chunks()` which I like better, thoughts about that?

I will also add some kind of DeprecationWarning to render_node/iter_node and then hopefully merge this PR in the next few days. Then we can hopefully also get #38 done. Feel free to have a look there too!

davepeck commented:
> I renamed stream_chunks() to iter_chunks() which I like better, thoughts about that?

I prefer `iter_chunks()` too!

> Then we can hopefully also get #38 done. Feel free to have a look there too!

Will do...

jodal commented:
I'm trying to get our version of `with_children` to work on components that also use contexts. I'm trying to juggle both the current implementation of htpy and these changes in my head at the same time, so that I won't have to redo it once this is merged.

What I'm saying is that I'd love for this to be merged and released, so that I can garbage collect half of my mental state :sweat_smile:

pelme commented:
have not been able to finish up these changes yet, not sure when I will get around to it. if someone wants to look into deprecation warnings for render_node/iter_node and update the docs/changelog, that would be helpful!

davepeck commented:
> have not been able to finish up these changes yet, not sure when I will get around to it. if someone wants to look into deprecation warnings for render_node/iter_node and update the docs/changelog, that would be helpful!

I took a quick first stab at this: https://github.com/davepeck/htpy/tree/pelme-zsxolror

Tried to match your house style but let me know what's most useful.

pelme commented:
I have updated the PR based on you commits @davepeck , I think it may be good to go.

I changed Renderable to be a type alias rather than a Protocol. I think that makes more sense. Then a Renderable can always be a Node.

@davepeck  @jodal Can you please have another look?

jodal commented on docs/changelog.md:
```diff
@@ -1,5 +1,9 @@
+- Add the `Renderable` protocol, a consistent API to render an `htpy` object as HTML or to iterate over it. `Element`, `Fragment`, `ContextProvider`, and `ContextConsumer` are all `Renderable`.
+- Deprecate `render_node()` and `iter_node()` and direct iteration over elements. Call `Renderable.__str__()` or `Renderable.iter_chunks()` instead. [Read the Usage docs for more details](usage.md#renderable).
+
 ## 25.4.0 - 2025-04-10
```
Missing #96 ;-)

pelme replied:
oops! will try to make sure changelogs are always part of PRs going forward so I do not forget. Fixed in main+updated docs+github release.

jodal commented on docs/static-typing.md:
```diff
@@ -47,6 +47,27 @@ def bootstrap_badge(
+
+htpy elements, fragments and context objects provides are "renderable". The `Renderable` type provides a consistent API to render a `htpy` object as HTML.
+
+The Renderable protocol defines these methods:
```
```suggestion
The `Renderable` protocol defines these methods:
```

jodal commented on docs/static-typing.md:
```diff
@@ -47,6 +47,27 @@ def bootstrap_badge(
+The Renderable protocol defines these methods:
+
+ - `.__str__()` - render as a HTML string by calling `str()`
+ - `.__html__()` render as a HTML string that is safe to use as markup. This makes it possible to directly embed a `Renderable` object in [Django/Jinja templates](django.md#using-htpy-as-part-of-an-existing-django-template).
```
```suggestion
 - `.__html__()` - render as a HTML string that is safe to use as markup. This makes it possible to directly embed a `Renderable` object in [Django/Jinja templates](django.md#using-htpy-as-part-of-an-existing-django-template).
```

pelme commented on htpy/__init__.py:
```diff
@@ -391,20 +442,12 @@ class _HasHtml(t.Protocol):
     def __html__(self) -> str: ...
 
 
+Renderable = BaseElement | ContextConsumer[t.Any] | ContextProvider[t.Any] | Fragment
```
Hmm, interesting!

I am not really against it but would also not like to make a backward compatibility commitment for `_iter_context()` at this point. What do you think about:

- Make `Renderable` a Protocol again which includes the `_iter_chunks_context()` method.
- Do not document or and explicitly state (with a comment in the Protocol definition) that `_iter_chunks()` is private and that there is no backward compatible guarantee.
- If you are indeed making your own Renderables, you need to know how the internals on htpy works and deal with potential changes.
- You and others can then play with the ability to implement arbitrary htpy objects.
- If this is successful and works out good and there are good use cases for this, we can make it a public API and document how you can extend htpy.

I updated the PR with this now, what do you think about that?

jodal replied:
Moving the core of the `_iter_chunks()` method out into each of the renderables is an obvious improvement once you see it, which I haven't until now. Love it, as this will give full control to any externally defined element as long as they get past the initial callable check in `_iter_chunks()`.

pelme commented on pyproject.toml:
```diff
@@ -4,7 +4,9 @@ description = "htpy - HTML in Python"
-    "markupsafe>=2.0.0"
+    "markupsafe>=2.0.0",
+    # typing_extensions is used for @typing.deprecated introduced in Python 3.13
+    "typing_extensions>=4.13.2",
```
good point! I also think it is kind of impossible to have a project without typing_extensions but it is nice to avoid it like that. updated the PR with your suggestion now.

pelme replied:
Congrats on having the PEP 750 accepted btw! Do you think we should add a `htpy.html()` function that turns t strings to htpy objects? We can continue that discussion in another issue.

jodal commented on htpy/__init__.py:
```diff
@@ -130,11 +135,26 @@ class ContextProvider(t.Generic[T]):
+    __html__ = __str__
+
+    def iter_chunks(self) -> Iterator[str]:
+        return _iter_chunks_node(self, {})
```
I guess we could skip a level in the call stack here and go directly for:

```suggestion
        return self._iter_chunks_context({})
```

Same for `ContextConsumer`, `BaseElement` and `Fragment`.

pelme replied:
very true, changed now!

we could also change `def iter_chunks(self) -> Iterator[str]:` to something like `def iter_chunks(self, context: dict[Context[t.Any], t.Any]=None) -> Iterator[str]:` and get rid of _iter_chunks_context altogether. then implementing Renderable would just need to implement public methods.

to make anything useful a custom implementation would still need to call into _iter_chunks_node (which we could rename to `iter_chunks_node` or something at that point. 

but maybe it is nicer to keep it as _iter_chunks_context for now and then see what other uses that comes up

pelme replied:
I went ahead and changed `iter_chunks()` to accept the context dict and got rid of `._iter_chunks_context()`. So now it is completely possible to implement anything outside of htpy by implementing `Renderable`. `_iter_chunks_node` is still private but easy to access via a fragment if one needs that from a custom Renderable.

davepeck reviewed (approved):
I'm really happy where this landed!
